### PR TITLE
Geometry renderer

### PR DESCRIPTION
**Pull request by barsoosayque**
#### Summary

SUMMARY: Infrastructure "Decouple rectangle rendering in SDL related code"

#### Purpose of change

First of all, decoupling and documenting the rendering code. Making it more modular and easier to tweak.

I'm working on a fix for I-Am-Erk/CDDA-Tilesets#133, more specifically: rendering a symbol in no sprite for item is found. I need fonts for that, and I can't use them since fonts are using hardcoded static variables and functions from `sdltiles.cpp`. To decouple fonts, I first need to decouple the mechanism to draw rectangles (since it actually uses two different implementations depending on options user selected).

#### Describe the solution

Introduce an interface which whole purpose is to render geometry. Use that interface instead of hardcoded functions.

Good:
- Non-intrusive
- Modular
- Easy to understand and use

Bad:
- vtables :(

#### Describe alternatives you've considered

Maybe using some template magic to reduce usage of virtual stuff. However, I don't think that is going to work, since that one option is known only at runtime, and this option decides what rectangle rendering mechanism to use.

#### Testing

Loaded up the game, everything works and renders. Color modulated option works. No performance impact is noticed.

### CODE REVIEW

File: src/sdl_geometry.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#pragma once
+#ifndef CATA_SRC_SDL_GEOMETRY_H
+#define CATA_SRC_SDL_GEOMETRY_H
+
+#if defined(TILES)
+#include "memory.h"
```
Comment (Qrox) [inline src/sdl_geometry.h:6]: This is neither a game header nor a standard c/c++ header.

File: src/sdl_utils.cpp [hunk 2]
```diff
@@ -82,13 +82,8 @@ SDL_Surface_Ptr create_surface_32( int w, int h )
...
 {
-    if( alt_rect_tex_enabled ) {
-        SetTextureColorMod( alt_rect_tex, r, g, b );
-        RenderCopy( renderer, alt_rect_tex, nullptr, &rect );
-    } else {
-        SetRenderDrawColor( renderer, r, g, b, 255 );
-        RenderFillRect( renderer, &rect );
-    }
+    SetRenderDrawColor( renderer, r, g, b, 255 );
+    RenderFillRect( renderer, &rect );
```
Comment (Qrox) [inline src/sdl_utils.cpp:86]: Why isn't this using the new renderer classes?

**Comment by Qrox** [comment]
> No performance impact is noticed.

Have you tested using debug menu - info - draw benchmark?

**Comment by barsoosayque** [comment]
> Have you tested using debug menu - info - draw benchmark?

I haven't :sweat: (Didn't know that was possible). Anyway, yes, the draw benchmark confirms that there is no performance decrease:

image

On the top is this PR, on the bottom -- master branch

File: src/sdl_geometry.h (hunk 1 above)
Comment (barsoosayque) [reply src/sdl_geometry.h:6]: Hol' up.. it is a [standard c++ header](https://en.cppreference.com/w/cpp/header/memory), right ? Or did you mean that I'm using quotes here and not angle brackets ?

File: src/sdl_utils.cpp (hunk 2 above)
Comment (barsoosayque) [reply src/sdl_utils.cpp:86]: I wasn't sure how to do it right. It was using a global method which in turn used global variables. 

Now that I'm lookin at references of this method, it is possible to pass geometry renderer into `cata_tiles`, and then into `pixel_minimap`, which would eliminate any need in this method at all.

File: src/sdl_geometry.h (hunk 1 above)
Comment (Qrox) [reply src/sdl_geometry.h:6]: The c++ header is `memory` (without the extension), not `memory.h`.

Comment (barsoosayque) [reply src/sdl_geometry.h:6]: Oh, I thought `<memory>` is just a shortcut for `memory.h` within the system defined include directories, but it's actually an implementation-defined feature. Good to know !

**Pull request merged**